### ISSUE TITLE
Invalid memory access in nfa_save_listids/nfa_restore_listids

### ISSUE THREAD

**Issue by andymass** [vim/vim#3175]
I'm trying to debug a difficult to reproduce crash, which happens with my configuration but not when reducing anything to find a minimal example.

While investigating I noticed that in regexp_nfa.c sometimes nfa_save_listids/nfa_restore_listids operates on the array listids which is too small.  I added a variable which stores the last allocated size and have seen prog->nstate become fairly large (around 900) while the buffer is only big enough for 5 or so ints.

I don't know if this behavior is exceptional and caused by some other bug but the logic here seems faulty somehow since the allocated size nstate has seemingly little to do with prog->nstate and there are no bound checks or reallocations.  

I'm hoping this admittedly vague description is enough to notice an error since I have not been able to come up with a reproducible example.

**Comment by dpelle** [comment]
Do you have a stack trace?
Can you reproduce it with valgrind?
Or can you reproduce it with the address sanitizer (i.e. rebuild vim after uncommenting the following line in vim/src/Makefile:
```
SANITIZER_CFLAGS = -g -O0 -fsanitize=address -fno-omit-frame-pointer
```
It it crashes, it will then provide useful information on stderr.

**Comment by andymass** [comment]
Thanks.  Address sanitizer: http://vpaste.net/IRY0P

Valgrind output: https://gist.github.com/andymass/0d4f7157ddec6d12666f419acd8b9597
Very noisy, but note "Invalid write of size 4" (could be red herring).

Backtrace: http://vpaste.net/szNC7

**Comment by dpelle** [comment]
@andymass wrote:

> Valgrind output: ...snip... Very noisy, but note "Invalid write of size 4"

Thanks. Regarding Valgrind noisy outout, it happens when using
the python interpreter in vim.  It's documented here as "expected
and not a problem" here:

https://svn.python.org/projects/python/trunk/Misc/README.valgrind

Anyway, valgrind also detects the same unexpected use of free
memory as ASAN in:
``` 
==23888== at 0x27132E: utf_ptr2char (mbyte.c:1793)
``` 
It's triggered by an autocommand.

Line number do not match what I see in the latest vim in git.
Which version of Vim did you use?  Ideally, you should try
with the latest version in git, in case bug has already been
fixed.

**Comment by brammool** [comment]
> Thanks.  Address sanitizer: http://vpaste.net/IRY0P
> 
> Valgrind output: https://gist.github.com/andymass/0d4f7157ddec6d12666f419acd8b9597
> Very noisy, but note "Invalid write of size 4" (could be red herring).
> 
> Backtrace: http://vpaste.net/szNC7

I can guess that the problem is that the number of states changes
between calls to recursive_regmatch().  I'll make a patch for that.

It would still be nice to have a pattern and text to reproduce this
with and have a test for it.  It's probably something with \& in the
pattern.

-- 
hundred-and-one symptoms of being an internet addict:
212. Your Internet group window has more icons than your Accessories window.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by andymass** [comment]
I've come up with a reproducible example (sanitizer output included): https://gist.github.com/andymass/4fc164de7e4cffe0272b56a1379da904

Unfortunately, it still crashes after Bram's update.  Maybe for a different reason.

@dpelle, sorry, the line numbers differed because of my own tracing code.  I am using master.

**Comment by brammool** [comment]
> I've come up with a reproducible example (sanitizer output included): https://gist.github.com/andymass/4fc164de7e4cffe0272b56a1379da904
> 
> Unfortunately, it still crashes after Bram's update.  Maybe for a different reason.
> 
> @dpelle, sorry, the line numbers differed because of my own tracing code.  I am using master.

I cannot reproduce the problem.

It's unlikely it is actually trying to use freed memory, more likely
pointer computations cause that memory location to be accessed.

Although the change for the listids is probably valid, it might be
completely unrelated to this problem, since it's happening when using
reginput.

-- 
If evolution theories are correct, humans will soon grow a third
hand for operating the mouse.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by andymass** [comment]
I think I see what is happening:

1) `searchpos()` is called with a complex pattern.
2) in `nfa_regexec_both()`, the global variable `nstate` is set to `prog->nstate` (large) then `nfa_regtry()`->`nfa_regmatch()` are called.
3) in `nfa_regmatch()`, vim does a `ui_breakcheck()` which eventually calls `clip_get_selection()`.
4) the `TextYankPost` autocmd is fired for the `"*` register which calls `match_file_pat()`->`vim_regexec()`->...->`nfa_regexec_both()`.  This sets `nstate` to `prog->nstate` for the autocmd filename pattern (small).
5) back in the context of (2), `nfa_regmatch()` calls `recursive_regmatch()` which calls `nfa_save_listids()`, which writes to invalid memory because `*listids` is only allocated according to the new global `nstate` set in (4) which is too small.

In short it is an autocmd's pattern interfering with the global variable `nstate` in the middle of another regex matching.

To reproduce it is necessary to either be running vim locally with X or with `ssh -Y` (trusted X forwarding) but `./configure && make` suffices.

**Comment by brammool** [comment]
Andy Massimino wrote:

> I think I see what is happening:
> 
> 1) `searchpos()` is called with a complex pattern.
> 2) in `nfa_regexec_both()`, the global variable `nstate` is set to `prog->nstate` (large) then `nfa_regtry()`->`nfa_regmatch()` are called.
> 3) in `nfa_regmatch()`, vim does a `ui_breakcheck()` which eventually calls `clip_get_selection()`.
> 4) the `TextYankPost` autocmd is fired for the `"*` register which calls `match_file_pat()`->`vim_regexec()`->...->`nfa_regexec_both()`.  This sets `nstate` to `prog->nstate` for the autocmd filename pattern (small).
> 5) back in the context of (2), `nfa_regmatch()` calls `recursive_regmatch()` which calls `nfa_save_listids()`, which writes to invalid memory because `*listids` is only allocated according to the new global `nstate` set in (4) which is too small.
> 
> In short it is an autocmd's pattern interfering with the global variable `nstate` in the middle of another regex matching.
> 
> To reproduce it is necessary to either be running vim locally with X or with `ssh -Y` (trusted X forwarding) but `./configure && make` suffices.

I'm glad you managed to dig this up!

We can either make sure regexec is never used recursively, or not use
TextYankPost for the clipboard.

That last one would be the simplest.  I wonder what people do with
TextYankPost anyway?  Is it relevant for the clipboard registers?

With some effort it would be possible to make regexec work recursively.
There is some support already (but I wonder when it works, since several
global values are not saved and restored).

-- 
You are Dead.  Do you wish to restart, load, or quit?

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by andymass** [comment]
Is it possible that other dangerous callbacks are processed in RealWaitForChar?  A cursory inspection shows there might be job, channel, resize, and title updates which can execute vim script.  If so, the problem still exists even if you were to remove the TextYankPost autocmd for the star register and these bugs will be even harder to reproduce in the future.

I'm not certain what a use case is for TextYankPost on the star register, but if this autocmd is the only possible issue, a simpler fix would be:

```
diff --git a/src/regexp_nfa.c b/src/regexp_nfa.c
index d72430fad..1787c4d39 100644
--- a/src/regexp_nfa.c
+++ b/src/regexp_nfa.c
@@ -7185,7 +7185,9 @@ nfa_regexec_both(
        prog->state[i].lastlist[1] = 0;
     }

+    block_autocmds();
     retval = nfa_regtry(prog, col, tm, timed_out);
+    unblock_autocmds();

     nfa_regengine.expr = NULL;
```

Although as mentioned this might not cover all the possible recursive cases.

> With some effort it would be possible to make regexec work recursively.

Could the remaining globals just be stuffed into `rex` and/or `*prog`?  

Another alternative would be to use a more restricted version of ui_breakcheck/RealWaitForChar which cannot perform unsafe callbacks during regex execution.

**Comment by brammool** [comment]
Please check if the problem still exists after patch 8.1.0192.
It should make executing a regexp recursively work.
But I haven't yet been able to trigger that in a test.


-- 
hundred-and-one symptoms of being an internet addict:
263. You have more e-mail addresses than shorts.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by andymass** [comment]
Looks like a good change.  I was able to get `nfa_regexec_both` to be entered recursively (modifying the source to printf when this occurs) with version 8.1.0194 without crashing using this example: https://gist.githubusercontent.com/andymass/8c9d570bb4445208302a317cfd421131/raw/4fe36a1c3cbe300398f2ab31ca5025004cdb1329/crash2.vim .  Crashes in vim 8.1.0191.
